enhancements/big-bundle-unpacking.md

```diff
@@ -0,0 +1,138 @@
+- a bundle data init container
+  - pull spec is used to delegate pulling of the bundle image via CRI-O
+  - mounts two volumes: bundle util and bundle data
+  - at runtime, executes the bundle tool from the bundle util mount to copy the bundle to the bundle data mount
+- a bundle extract container
+  - contains opm
+  - mounts the bundle data volume
+  - runs `opm extract` against the bundle data mount which results in a `ConfigMap` being created that contains the bundle’s kubectl-able content
+
+#### Partitioning and Storage
+
+Instead of extracting bundle content directly to a `ConfigMap` -- as outlined in the [previous section](#podspec-pull-requirement) -- the unpack `Job` will first encode it as JSON and compress the result. The compressed JSON will then be split into chunks of a [fixed size](#determining-chunk-size), each wrapped with a header to keep track of its position in the chunk sequence. Each chunk is then encoded as JSON and stored in a respective `ConfigMap`. To each of these `ConfigMaps` a `porcelain.operators.coreos.com/bundle: <bundle-name>` label is applied to faciliate later retrieval. A `porcelain.operators.coreos.com/signature: <signature>` annotation is also generated for each chunk and is used to validate its integrity. When the content for a particular bundle is then retrieved, the `ConfigMaps` with the matching bundle name label are collected, sorted by the position tracked in each header, reconstituted into a single artifact, and decompressed before being returned for a `/content` request. If the signature verification fails for any chunk, a status condition indicating will be added to the associated bundle which will trigger a new unpack `Job` to run in a bid to rectify the issue.
+
+#### Determining Chunk Size
+
+Chunk-size will be determined by an option on the `catalog-operator` command, which if unset, directs the operator to probe the cluster for the resource storage limit. One potential probing strategy could be to create ConfigMaps of various sizes; akin to a "binary search" for the limit.
+
+### Risks and Mitigations
+
+Decompressing bundle content for each request may add significant time complexity. Some options for mitigating this are:
+
+- building an [MRU cache](https://en.wikipedia.org/wiki/Cache_replacement_policies#Most_recently_used_(MRU)) of bundle content
+- avoiding content compression entirely
+
+## Design Details
+
+### Version Skew Strategy
+
+#### The `Bundle` API and OLM
+
+The default aggregated API-server to service `Bundle` requests will be compiled and shipped w/ the rest of OLM; i.e. it will always match the version of OLM running on a given cluster.
+
+#### Unpack Jobs and OPM
+
+A small, statically linked binary will be generated from in-tree OLM source, for use in the unpack `Job`. This binary will vendor OPM packages to tightly couple content extraction logic to OLM version.
+
+#### Extraction Logic and Operator Bundle Mediatypes
```

Comment:
If the operator is removed/upgraded, would the corresponding configmaps be cleaned up as well? Are there any ownerrefs on the configmaps?

Reply:
I guess we can keep the same logic that we have in-place today.

Reply:
I omitted specifics on the cleanup b/c I figured that wasn't important to grasp the idea I was trying to outline.

I think there should be ownerreferences between sibling segment configmaps as well as an option in the unpacker to automatically garbage collect any resource w/ a segment that fails an integrity check.

Reply:
Somewhat related question: what's the lifetime of the Bundle object? Is it created on demand when an InstallPlan with a corresponding bundleLookup is reconciled? Does it make sense to have the InstallPlan own the Bundle, so that when the InstallPlan succeeds (and gets removed?), the Bundle (and its ConfigMaps) gets removed too?